tests/generators/test_cpp.py

```diff
@@ -0,0 +1,668 @@
+        a_capturing_logger,
+        tmp_path_factory,
+    ):
+        output_dir = tmp_path_factory.mktemp("test_cpp_generate_warnings")
+        cpp_generator.generate(
+            declaration_exotic_values, {}, output_dir, lambda _: _, a_capturing_logger
+        )
+
+        assert a_capturing_logger.handlers[0].messages == warnings_exotic_values
+
+    @pytest.mark.parametrize(
+        "id_type",
+        ("package_name", "specification_namespace", "trait_namespace"),
+    )
+    @pytest.mark.parametrize(
+        "reserved_word",
+        (
+            # Check a representative keyword.
+            next(iter(cpp_keywords.keywords)),
+            "__dunder_implicitly_reserved",
+            "traits",
+            "specifications",
+        ),
+    )
+    def test_when_identifiers_reserved_keywords_then_exception_raised(
+        self,
+        tmp_path_factory,
+        a_capturing_logger,
+        declaration_invalid_identifiers,
+        id_type,
+        reserved_word,
+    ):
+        output_dir = tmp_path_factory.mktemp("test_cpp_invalid_identifiers")
+        with pytest.raises(ValueError) as err:
+            cpp_generator.generate(
+                declaration_invalid_identifiers(**{id_type: reserved_word}),
+                {},
+                output_dir,
+                lambda _: _,
+                a_capturing_logger,
+            )
+        assert (
+            str(err.value)
+            == f"'{reserved_word}' (from '{reserved_word}') is a reserved keyword."
+        )
+
+
+def test_cpp_project(generated_path, tmp_path_factory, cpp_project_dir):
+    build_dir = tmp_path_factory.mktemp("test_cpp_project")
+
+    # Install third-party C++ library dependencies.
+    subprocess.check_call(
+        [
+            "conan",
+            "install",
+            "--install-folder",
+            f"{build_dir}/.conan",
+            "-g",
+            "CMakeDeps",
+            "-s",
+            "build_type=RelWithDebInfo",
+            cpp_project_dir,
+        ]
+    )
+
+    include_paths = ";".join(
+        (
+            f"{generated_path}/openassetio_traitgen_test_all/include",
+            f"{generated_path}/openassetio_traitgen_test_specifications_only/include",
+            f"{generated_path}/openassetio_traitgen_test_traits_only/include",
+        )
+    )
+
+    # Configure CMake project
+    subprocess.check_call(
+        [
+            "cmake",
+            "-S",
+            cpp_project_dir,
+            "-B",
+            build_dir,
+            "--preset",
+            "test",
+            f"-DCMAKE_PREFIX_PATH={build_dir}/.conan",
+            f"-DOPENASSETIO_TRAITGENTEST_ADDITIONAL_INCLUDE_DIRS={include_paths}",
+        ]
+    )
+    subprocess.check_call(["cmake", "--build", build_dir])
+    # Run tests in CMake project.
+    subprocess.check_call(["ctest", "-VV", "--test-dir", build_dir])
+
+
+#
+# Fixtures
+#
+
+
+@pytest.fixture
+def docstring_for(rootnode_for, cpp_language):
```

Comment:
[fix] Can we document these methods, and a high-level overview of how it works?

Reply:
Added a couple of docstrings in c36c59a6